### DISCUSSION TITLE
How to handle typedefs

### DISCUSSION THREAD

**Discussion by nyholku** [bytedeco/javacpp#541] [Q&A] [answered]
Hi,

I've been experimenting with OCC headers to find all the potentials issues in creating the bindings with JavaCPP.

OCC defines:

```
#define Handle(Class) opencascade::handle<Class>
#define DEFINE_STANDARD_HANDLECLASS(C1,C2,BC) class C1; typedef Handle(C1) Handle_##C1;
#define DEFINE_STANDARD_HANDLE(C1,C2) DEFINE_STANDARD_HANDLECLASS(C1,C2,Standard_Transient)
```
I can now get this C++ to work (ie Java/JNI wrapping generated and compiled and apparently also executing correctly):

```
opencascade::handle(SomeClass) variable;
```

this also works:

```
Handle(SomeClass) variable;
```

but this does not work:

```
DEFINE_STANDARD_HANDLE(SomeClass,Standard_Transient)

Handle_SomeClass variable;
```

This results in:

```
occ/TKernel.java:1348: error: cannot find symbol
public static native @MySmartPtr Base someFoo(@ByVal Handle_Base base);
                                                     ^
  symbol:   class Handle_Base
  location: class TKernel
```

Now, this is not necessarily a problem as most of the OCC API seems to use `Handle(SomeClass)`
but this raises the question how to handle typedefs?

**Comment by saudet** [comment, marked as answer]
This isn't working because of a typedef, it's because the template instance is missing. Those need to be given a name:
https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes#creating-instances-of-c-templates
In this case, something like this should work:
```java
infoMap.put(new Info("opencascade::handle<Base>").pointerTypes("Handle_Base"));
infoMap.put(new Info("opencascade::handle<SomeClass>").pointerTypes("Handle_SomeClass"));
...
```
You can put this in methods and for loops like this to cut down on the copy/paste:
https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/src/main/java/org/bytedeco/pytorch/presets/torch.java#L1232
https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/src/main/java/org/bytedeco/pytorch/presets/torch.java#L511
https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/src/main/java/org/bytedeco/pytorch/presets/torch.java#L1283

**Reply by nyholku** [reply]
Thanks, I see, I think ;)
